src/main/java/org/spongepowered/api/item/inventory/menu/KeySwapHandler.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.item.inventory.menu;
+
+import org.spongepowered.api.item.inventory.Container;
+import org.spongepowered.api.item.inventory.Slot;
+import org.spongepowered.api.item.inventory.slot.SlotIndex;
+
+@FunctionalInterface
+public interface KeySwapHandler {
+
+    /**
+     * Handler for inventory {@link ClickTypes#KEY_SWAP} callbacks.
+     *
+     * @param container the container
+     * @param slot the primary clicked slot
+     * @param slotIndex the primary clicked slot index
+     * @param clickType the type of click
+     * @param slot2 the secondary slot in the interaction.
+     *
+     * @return false to prevent the action
+     */
+    boolean handle(Container container, Slot slot, SlotIndex slotIndex, ClickType clickType, Slot slot2);
```

Comment:
Why? Imo that would just be another additional interface for every handler added to the API without much benefit.

Reply:
This change will avoid long lambda parameter list.
You even can remove these interfaces and use something like `Predicate<KeySwapHandlerData>` in `handleXXX` methods.
`Predicate`, `Function`, `Consumer` everything you want.

Reply:
You don't have to use lambdas. You can just pass an instance of smth. that implements all those interfaces or method reference that fits.
Still don't see the benefit of wrapping the data to just unwrap it in the handler again when you want to use it.

Reply:
This would give shorter lambda. Compare:
```java
KeySwapHandler = (container, slot, slotIndex, clickType, slot2) => {
    if (slot2 != null) {
        s.peek();
    }
};
```
and this:
```java
Consumer<KeySwapData> = data => {
  data.getSlot2().ifPresent(Slot::peek);
};
```

Reply:
Also, my approach is more deprecation/addition safer way.
You could deprecate getter methods and add new getter without breaking anything.

Reply:
The original idea of the callbacks was to reduce the massive overhead we currently have with inventory events. 
We have to detect and collect all changes and then based on cancel state rollback everything

If you just want to have some kind of immutable menu its easier to just have callbacks for clicks/changes instead.

While your approach would make deprecation/addition easier it also would make it impossible to just implement all those interfaces in one class.
Example here: https://github.com/SpongePowered/SpongeCommon/pull/2211/files#diff-a4229ef51d31c0aab92651c1562a8ab9R435
And let's be honest, we would find a way to break it anyways.

I'm still not convinced that a wrapper around those values is the best way to go.